app/src/main/java/de/xikolo/controllers/course/CourseActivity.kt

```diff
@@ -0,0 +1,312 @@
+        enrollBar?.visibility = View.VISIBLE
+        enrollButton?.isEnabled = false
+        enrollButton?.isClickable = false
+        enrollButton?.setText(R.string.btn_starts_soon)
+    }
+
+    override fun finishActivity() {
+        finish()
+    }
+
+    override fun restartActivity() {
+        finish()
+        startActivity(intent)
+    }
+
+    override fun showProgressDialog() {
+        if (progressDialog == null) {
+            progressDialog = ProgressDialogAutoBundle.builder().build()
+        }
+        progressDialog?.show(supportFragmentManager, ProgressDialog.TAG)
+    }
+
+    override fun hideProgressDialog() {
+        if (progressDialog?.dialog?.isShowing == true) {
+            progressDialog?.dismiss()
+        }
+    }
+
+    override fun showErrorToast() {
+        ToastUtil.show(R.string.error)
+    }
+
+    override fun showNoNetworkToast() {
+        ToastUtil.show(R.string.toast_no_network)
+    }
+
+    override fun showLoginRequiredMessage() {
+        ToastUtil.show(R.string.toast_please_log_in)
+    }
+
+    override fun openLogin() {
+        val intent = LoginActivityAutoBundle.builder().build(this)
+        startActivity(intent)
+    }
+
+    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
+    override fun onNetworkEvent(event: NetworkStateEvent) {
+        super.onNetworkEvent(event)
+
+        if (event.isOnline) {
+            toolbar.subtitle = ""
+            tabLayout.setBackgroundColor(resources.getColor(R.color.apptheme_toolbar))
+            setColorScheme(R.color.apptheme_toolbar, R.color.apptheme_statusbar)
+        } else {
+            toolbar.subtitle = getString(R.string.offline_mode)
+            tabLayout.setBackgroundColor(resources.getColor(R.color.offline_mode_toolbar))
+            setColorScheme(R.color.offline_mode_toolbar, R.color.offline_mode_statusbar)
+        }
+    }
+
+    override fun getPresenterFactory(): PresenterFactory<CoursePresenter> {
+        return CoursePresenterFactory()
+    }
+
+    inner class CoursePagerAdapter internal constructor(private val fragmentManager: FragmentManager) : FragmentPagerAdapter(fragmentManager), TabLayout.OnTabSelectedListener {
+
+        override fun getPageTitle(position: Int): CharSequence? {
+            return getString(areaState.get(position).titleRes)
+        }
+
+        override fun getCount(): Int {
+            return areaState.size
+        }
+
+        override fun getItem(position: Int): Fragment? {
+            // Check if this Fragment already exists.
+            // Fragment Name is saved by FragmentPagerAdapter implementation.
+            val name = makeFragmentName(R.id.viewpager, position)
+            var fragment: Fragment? = fragmentManager.findFragmentByTag(name)
+            if (fragment == null && courseId != null) {
+                when (areaState.get(position)) {
+                    CourseArea.LEARNINGS      -> fragment = LearningsFragmentAutoBundle.builder(courseId!!).build()
```

Comment:
Kotlin's auto unwrap for `courseId!!` doesn't work here with the if statement? So the `!!` is still needed?

Reply:
Because it is mutable, it could have been changed at the time of access. I made a `let` block with it in it.